monty-31/src/x86_64_avx512/utils.rs

```diff
@@ -0,0 +1,221 @@
+use core::arch::x86_64::{self, __m512i};
+use core::mem::transmute;
+
+use crate::{MontyParameters, PackedMontyParameters, TwoAdicData};
+
+// Godbolt file showing that these all compile to the expected instructions. (Potentially plus a few memory ops):
+// https://godbolt.org/z/dvW7r1zjj
+
+/// Halve a vector of Monty31 field elements in canonical form.
+/// If the inputs are not in canonical form, the result is undefined.
+#[inline(always)]
+pub fn halve_avx512<MP: MontyParameters>(input: __m512i) -> __m512i {
+    /*
+        We want this to compile to:
+            vptestmd least_bit, val, ONE
+            vpsrld   t, val, 1
+            vpaddd   res, least_bit, t, maybe_half
```

Comment:
This is the correct syntax for masks: (plus i'm adding an extra space for visual clarity)
```suggestion
            vptestmd  least_bit, val, ONE
            vpsrld    res, val, 1
            vpaddd    res{least_bit}, res, maybe_half
```